src/zasm.zig

```diff
@@ -0,0 +1,716 @@
+    }
+
+    /// Creates a new physical address.
+    ///
+    /// ## Panics
+    /// This function panics if a bit in the range 52 to 64 is set.
+    pub fn initPanic(addr: u64) PhysAddr {
+        return init(addr) catch @panic("physical addresses must not have any bits in the range 52 to 64 set");
+    }
+
+    const TRUNCATE_CONST: u64 = 1 << 52;
+
+    /// Creates a new physical address, throwing bits 52..64 away.
+    pub fn initTruncate(addr: u64) PhysAddr {
+        return PhysAddr{ .value = addr % TRUNCATE_CONST };
+    }
+
+    /// Creates a new physical address, without any checks.
+    pub fn initUnchecked(addr: u64) PhysAddr {
+        return .{ .value = addr };
+    }
+
+    /// Creates a physical address that points to `0`.
+    pub fn zero() PhysAddr {
+        return .{ .value = 0 };
+    }
+
+    /// Convenience method for checking if a physical address is null.
+    pub fn isNull(self: PhysAddr) bool {
+        return self.value == 0;
+    }
+
+    /// Aligns the physical address upwards to the given alignment.
+    /// The alignment must be a power of 2 and greater than 0.
+    pub fn alignUp(self: PhysAddr, alignment: usize) PhysAddr {
+        return .{ .value = std.mem.alignForward(self.value, alignment) };
+    }
+
+    /// Aligns the physical address downwards to the given alignment.
+    /// The alignment must be a power of 2 and greater than 0.
+    pub fn alignDown(self: PhysAddr, alignment: usize) PhysAddr {
+        return .{ .value = std.mem.alignBackward(@intCast(usize, self.value), alignment) };
+    }
+
+    /// Checks whether the physical address has the given alignment.
+    /// The alignment must be a power of 2 and greater than 0.
+    pub fn isAligned(self: PhysAddr, alignment: usize) bool {
+        return std.mem.isAligned(self.value, alignment);
+    }
+
+    test {
+        std.testing.refAllDecls(@This());
+        try std.testing.expectEqual(@bitSizeOf(u64), @bitSizeOf(PhysAddr));
+        try std.testing.expectEqual(@sizeOf(u64), @sizeOf(PhysAddr));
+    }
+};
+
+// ~~~ [ zig-x86_64/src/stuctures/paging/page.zig ] ~~~
+
+const SIZE_4KiB_STR: []const u8 = "4KiB";
+const SIZE_2MiB_STR: []const u8 = "2MiB";
+const SIZE_1GiB_STR: []const u8 = "1GiB";
+
+pub const PageSize = enum {
+    Size4KiB,
+    Size2MiB,
+    Size1GiB,
+
+    pub fn bytes(self: PageSize) u64 {
+        return switch (self) {
+            .Size4KiB => 4096,
+            .Size2MiB => 4096 * 512,
+            .Size1GiB => 4096 * 512 * 512,
+        };
+    }
+
+    pub fn sizeString(self: PageSize) []const u8 {
+        return switch (self) {
+            .Size4KiB => SIZE_4KiB_STR,
+            .Size2MiB => SIZE_2MiB_STR,
+            .Size1GiB => SIZE_1GiB_STR,
+        };
+    }
+
+    pub fn isGiantPage(self: PageSize) bool {
+        return self == .Size1GiB;
+    }
+};
+
+// ~~~ [ zig-x86_64/src/instructions/tlb.zig ] ~~~
+
+/// Structure of a PCID. A PCID has to be <= 4096 for x86_64.
+pub const Pcid = packed struct {
```

Comment:
Doesn't this seem overly boilerplate:y / overly verbose? Can't we just have a type synonym like:

```
PcidType = u12
```

Reply:
I agree. This feels like unnecessary boiler plate. It would make sense if they added methods to Pcid.

We can do an alias like this:

```zig
const PcidType = u12
```

However, I still propose that we keep the code as is, as it makes maintenance easier when we update the code against the `x86_64` repo. Make minimal changes unless warranted. And in this case, the "make it slightly cleaner", vs. "simplify maintenance" does not seem worth it.

However, I can be persuaded of course :)